Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm

```diff
@@ -37,121 +37,205 @@
+}
+
+void NetworkTransportSession::setupListenerConnectionHandler()
+{
+    nw_listener_set_new_connection_handler(m_listener.get(), makeBlockPtr([weakThis = WeakPtr { *this }] (nw_connection_t inboundConnection) mutable {
+        ASSERT_UNUSED(weakThis, weakThis);
+        // FIXME: Implement incoming connection handler
+    }).get());
+}
+
+RetainPtr<nw_parameters_t> NetworkTransportSession::createParameters(NetworkTransportStreamType streamType)
+{
+    ASSERT(streamType != NetworkTransportStreamType::IncomingUnidirectional);
+    RetainPtr<nw_parameters_t> parameters = adoptNS(nw_parameters_copy(nw_connection_copy_parameters(m_connections.first().get())));
+    if (!parameters) {
+        ASSERT_NOT_REACHED();
+        return nil;
+    }
+    nw_protocol_stack_t default_stack = nw_parameters_copy_default_protocol_stack(parameters.get());
+    nw_protocol_stack_iterate_application_protocols(default_stack, [&streamType](nw_protocol_options_t options) {
+        if (nw_protocol_options_is_webtransport(options)) {
+            nw_webtransport_options_set_is_unidirectional(options, streamType == NetworkTransportStreamType::OutgoingUnidirectional);
+            nw_webtransport_options_set_is_datagram(options, false);
+        }
+    });
+    return parameters;
+}
+
+void NetworkTransportSession::receiveDatagramLoop()
+{
+    nw_connection_receive(m_connections.first().get(), 1, std::numeric_limits<uint32_t>::max(), makeBlockPtr([weakThis = WeakPtr { *this }] (dispatch_data_t content, nw_content_context_t, bool withFin, nw_error_t error) {
+        RefPtr strongThis = weakThis.get();
+        if (!strongThis)
+            return;
+        if (error)
+            return; // FIXME: Pipe this error to JS.
+
+        // FIXME: Not only is this an unnecessary string copy, but it's also something that should probably be in WTF or FragmentedSharedBuffer.
+        auto vectorFromData = [](dispatch_data_t content) {
+            ASSERT(content);
+            Vector<uint8_t> request;
+            dispatch_data_apply_span(content, [&](std::span<const uint8_t> buffer) {
+                request.append(buffer);
+                return true;
+            });
+            return request;
+        };
+
+        strongThis->receiveDatagram(vectorFromData(content).span());
+        strongThis->receiveDatagramLoop();
+    }).get());
+}
+
+void NetworkTransportSession::createStream(NetworkTransportStreamType streamType, CompletionHandler<void(std::optional<WebTransportStreamIdentifier>)>&& completionHandler)
+{
+    ASSERT(streamType != NetworkTransportStreamType::IncomingUnidirectional);
+    RetainPtr<nw_parameters_t> parameters = createParameters(streamType);
+    if (!parameters) {
+        ASSERT_NOT_REACHED();
+        return completionHandler(std::nullopt);
+    }
+
+    RetainPtr connection = nw_connection_create_with_connection(m_connections.first().get(), parameters.get());
```

Comment:
This needs adoptNS, and forward declaration in the SPI header.  Is there no API for this?

Reply:
Replaced with connection group.